Posts/2021/02/XML-NewEmployees.md

```diff
@@ -0,0 +1,180 @@
+    
+    if (Test-Path $XMLPath) {
+        
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
+    }
+    else {
+        Start-Process "$PSScriptRoot\Write-XML.ps1" -Wait
+        $XML = New-Object -TypeName XML
+        $XML.Load($XMLPath)
+    }
+    
+```
+Here, we are testing the path for the XML and, if it doesn't exist, running the creation script again. Then, we're loading it into a variable for reference.
+
+We can gather information for the new users by making use of "Read-Host" and some added validation. Like so:
+
+```powershell
+[ValidatePattern("\w+")]$FirstName = Read-Host -Prompt "Please input users first name"
+[ValidatePattern("\w+")]$LastName = Read-Host -Prompt "Please input users last name"
+[ValidatePattern("\w+")]$Company = Read-Host -Prompt "Please input users company"
+[ValidatePattern("\w+")]$Office = Read-Host -Prompt "Please input users office"
+```
+Dead simple. 4 questions with 4 answers, into 4 variables. Instead of asking lots of questions to find out information like postcode or country, let's use our XML!
+
+```powershell
+# Declare some more variables
+$FirstInitial = "$FirstName.substring(0,1)"
+$Path = 'OU=Users,OU=Contoso,DC=Contoso,DC=com'
+
+# Build hashtable of new user parameters
+  $NewUserParams =@{
+    
+        GivenName = $FirstName
+        Name = "$FirstName $LastName"
+        Surname = $LastName
+        SamAccountName = "$FirstInitial + $LastName"
+        DisplayName = "$FirstName $LastName"
+        UserPrincipalName = "$userName" + "@" + "contoso.com"
+        AccountPassword = $password
+        Office = $Office
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+        Path = $Path
+        ChangePasswordAtLogon = $false
+        Enabled = $true
+    }
+    
+# Create the user
+New-ADUser @NewUserParams
+```
+Simple as that, browsing through the xml elements using dot-notation, and trimming off any unwanted whitespace. This is cropped below for convenience:
+
+```powershell
+
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+```
+# Helpful Notes
+* Be careful when creating your elements! No whitespace allowed!
+* Your inputted variables using Read-Host need to match the references in the XML. You could make use of Validation Scripts to validate your inputs. Here is a helpful read on everything to do with validating inputs: https://adamtheautomator.com/powershell-validatescript/
```

Comment:
change the start of the sentence to "The variables you input using `Read-Host` need to match"

Also, fix URL  "... Here is a helpful read on [everything to do with validation inputs.][https://adamtheautomator.com/powershell-validatescript/]"